h/migrations/versions/7bf167611bc3_add_the_groupscope_table.py

```diff
@@ -0,0 +1,39 @@
+"""
+Add the GroupScope table.
+
+Revision ID: 7bf167611bc3
+Revises: c943c3f8a7e5
+Create Date: 2018-02-08 11:00:50.420618
+"""
+
+from __future__ import unicode_literals
+
+from alembic import op
+import sqlalchemy as sa
+
+revision = '7bf167611bc3'
+down_revision = 'c943c3f8a7e5'
+
+
+def upgrade():
+    op.create_table(
+        'groupscope',
+        sa.Column('id', sa.Integer, autoincrement=True, primary_key=True),
+        sa.Column('hostname', sa.UnicodeText, nullable=False, unique=True),
+    )
+    op.create_table(
+        'group_groupscope',
+        sa.Column(
+            'group_id', sa.Integer, sa.ForeignKey('group.id'), nullable=False),
+        sa.Column(
+            'groupscope_id',
+            sa.Integer,
+            sa.ForeignKey('groupscope.id'),
+            nullable=False),
+        sa.PrimaryKeyConstraint('group_id', 'groupscope_id'),
+    )
```

Comment:
What is the deletion behaviour here if you remove a group? I think this will error out if there are any scopes for a group rather than just cascading the deletion, which I think is the behaviour we want.

Reply:
If talking to pg directly I think you'd have to:

1. Delete the matching entries from the group_groupscope relation table
2. Delete the group, or delete the groupscope if you were trying to delete a groupscope

There's no direct connection between the group and groupscope tables, so you don't need to delete groupscopes before you can delete a group or vice-versa, but you do need to delete entries in the relation table before you can delete either a group or a groupscope.

`ON DELETE CASCADE` could be used to do this automatically if we wanted.

I think sqlalchemy would do it automatically already too.

Also, deleting group and/or group_groupscope rows will not automatically delete groupscopes. So it's possible to end up with unused groupscopes left in the DB, not related to any groups. I'm not sure there's anything we can do about this in DDL though.

Reply:
Again, I got a bit confused for a moment forgetting that scopes exist independently of groups. I do think it would make sense to cascade deletions from the `group_groupscope` table when groups are deleted though. I was fairly sure this could be set up using SQLAlchemy.

Reply:
I had a _quick_ look into deletion behaviour here. At the PostgreSQL level the short answer is:

* You have to delete the `group_groupscope` entries before you can delete either a `group` or a `groupscope`. This means that it's not possible for orphaned `group_groupscope` rows to be left hanging around in the DB. Not because they're deleted automatically, but because you have to delete them manually first. This is because a `group_groupscope` row contains not-nullable foreign keys to both `group` and `groupscope`.

* There's nothing stopping you from deleting the `group_groupscope` rows but not deleting the `groupscope`(s). So it _is_ possible to have orphaned `groupscope` rows left in the DB, at least if someone has been editing the DB in SQL directly.

  I think there's probably something that can be done about this at the PostgreSQL level by registering a function that's triggered on delete of a `group_groupscope` and automatically deletes any orphaned `groupscope`s. But I'm not sure it's worth the trouble of going as far as Postgres trigger functions especially since we're using Postgres via SQLAlchemy and would have to figure out how to do that.

* Because it's a many-many relationship between `groupscope` and `group_groupscope` I don't think normal `ON DELETE CASCADE`-type stuff can help you here. You would need a trigger function.

  This is because `groupscope` does not contain a foreign key referencing `group_groupscope`. If it did, you could put an `ON DELETE CASCADE` on that foreign key and deleting the `group_groupscope` would automatically delete the `groupscope`. But because a single `groupscope` can be related to many `group_groupscope`s, the foreign key is actually on `group_groupscope` not on `groupscope`.

Details:

```
INSERT INTO public.group(pubid, authority, name) VALUES ('abc123', 'localhost', 'mygroup');
-> Creates a group with id 1.

INSERT INTO groupscope(origin) VALUES ('https://foo.example.com:5000');
-> Creates a groupscope with id 1.

INSERT INTO group_groupscope(group_id, groupscope_id) VALUES (1, 1);

postgres=# DELETE FROM public.group;                      
ERROR:  update or delete on table "group" violates foreign key constraint "fk__group_groupscope__group_id__group" on table "group_groupscope"
DETAIL:  Key (id)=(1) is still referenced from table "group_groupscope".

postgres=# DELETE FROM groupscope ;
ERROR:  update or delete on table "groupscope" violates foreign key constraint "fk__group_groupscope__groupscope_id__groupscope" on table "group_groupscope"
DETAIL:  Key (id)=(1) is still referenced from table "group_groupscope".

postgres=# DELETE FROM group_groupscope;
-> The two previous DELETEs above would work now.
```

Reply:
^ Following some [Slack discussion](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1518177230000008) I think this pr has modelled the relation between group and groupscope all wrong, gonna close this and submit a new one.